File: frequencyTopic2020/list/findMedianSortedArrays.cc
#include <iostream>
#include <vector>

using namespace std;

/*
// burst force by merge arr
class Solution {
 public:
  double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
    int m = nums1.size();
    int n = nums2.size();
    if (m + n == 0) {
      return 0;
    }
    vector<int> merge = vector<int>(m + n, 0);
    int p = 0;
    int i = 0;
    int j = 0;
    while (i < m && j < n) {
      if (nums1[i] <= nums2[j]) {
        merge[p] = nums1[i];
        i++;
      } else {
        merge[p] = nums2[j];
        j++;
      }
      p++;
    }
    if (i < m) {
      while (i < m) {
        merge[p] = nums1[i];
        i++;
        p++;
      }
    } else if (j < n) {
      while (j < n) {
        merge[p] = nums2[j];
        j++;
        p++;
      }
    }
    if ((m + n) % 2 == 1) {
      return merge[(m + n) / 2];
    }
    return (merge[(m + n) / 2] + merge[(m + n) / 2 + 1]) / 2.0;
  }
};
*/

class Solution {
 public:
  double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {

    if (nums2.size() > nums1.size()) {
      return findMedianSortedArrays(nums2, nums1);
    }
    int m = nums1.size();
    int n = nums2.size();

    int nLeftPart = (m + n + 1) / 2;
    int left = 0;
    int right = m;
    int m1 = 0;
    int m2 = 0;

    while (left <= right) {
      int mid = (left + right) / 2;
      int nid = nLeftPart - mid;

      int leftM = (mid == 0 ? INT_MIN : nums1[mid - 1]);
      int rightM = (mid == m ? INT_MAX : nums1[mid]);
      int leftN = (nid == 0 ? INT_MIN : nums2[nid - 1]);
      int rightN = (nid == n ? INT_MAX : nums2[nid]);
      if (leftM <= rightN) {
        left = mid + 1;
        m1 = max(leftM, leftN);
        m2 = min(rightM, rightN);
      } else {
        right = mid - 1;
      }
    }

    if ((m + n) % 2 == 1) {
      return m1;
    }
    return (m1 + m2) / 2.0;
  }
};

File: rush/arrStr/findMedianSortedArrays.go
package main

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {

    if len(nums1) == 0 && len(nums2) == 0 {
        return 0
    }
    if len(nums1) > len(nums2) {
        return findMedianSortedArrays(nums2, nums1)
    }

    total := (len(nums1) + len(nums2) + 1) / 2
    left, right := 0, len(nums1)
    for left < right {
        mid := (left + right + 1) / 2
        p := total - mid
        if nums1[mid - 1] > nums2[p] {
            right = mid - 1
        } else  {
            left = mid
        }
    }

    var n1left, n1right, n2left, n2right int
    p1, p2 := left, total - left
    intMin, intMax := math.MinInt64, math.MaxInt64
    if p1 == 0 {
        n1left = intMin
    } else {
        n1left = nums1[p1 - 1]
    }
    if p2 == 0 {
        n2left = intMin
    } else {
        n2left = nums2[p2 - 1]
    }

    if p1 == len(nums1) {
        n1right = intMax
    } else {
        n2right = nums1[p1]
    }
    if p2 == len(nums2) {
        n2right = intMax
    } else {
        n2right = nums2[p2]
    }

    fmt.Println("x", n1left, n1right, n2left, n2right)

    if (len(nums1) + len(nums2)) % 2 == 0 {
        return float64(min(n1right, n2right) + max(n1left, n2left)) / 2.0
    }
    return float64(max(n1left, n2left))
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}
